### PR TITLE
Blog post: Changes to Rust's WASI targets

### PR DESCRIPTION
**Pull request by yoshuawuyts**
This explains the upcoming changes to Rust's WASI targets, starting in Rust 1.78. This supersedes #1099. I put the post date in for 2024-04-10 (two days from now). Thanks!

cc/ @alexcrichton, @wesleywiser

### CODE REVIEW

File: posts/2024-04-10-updates-to-rusts-wasi-targets.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+"WASI, snapshot 1" target as `wasm32-wasi`. We should have instead chosen to add
+some suffix to the initial target triple so that the eventual stable WASI 1.0
+target can just be called `wasm32-wasi`.
+
+In anticipation of both an eventual WASI 1.0 target, and to preserve consistency
+between target names, we'll begin rolling out a name change to the existing WASI
+0.1 target. Starting in Rust 1.78 (May 2nd, 2024) a new `wasm32-wasip1` target
+will become available. Starting Rust 1.81 we will begin warning existing users
+of `wasm32-wasi` to migrate to `wasm32-wasip1`. And finally in Rust 1.84
+(January 9th, 2025) the `wasm32-wasi` target will no shipped on the stable
```
Comment (CryZe) [inline posts/2024-04-10-updates-to-rusts-wasi-targets.md:60]:
```suggestion
(January 9th, 2025) the `wasm32-wasi` target will no longer be shipped on the stable
```

**Comment by yoshuawuyts** [comment]
Talked to @alexcrichton over Zulip today, and he said it looks good. I shared an earlier draft with him last week, and the version I posted here already incorporated all of his feedback. He's currently traveling though, and can't directly approve it until he's back - hence this update.

**Review by wesleywiser — APPROVED** [review]

**Review by rylev — COMMENTED** [review]
I think we're burying the lead a bit here. This sentence might make readers wonder about why this post is on the main Rust blog instead of Inside Rust or even somewhere else. They may ask themselves "why should I read a post about a target I shouldn't use yet?".

The main story here is that there is a shift in target naming coming that will require user action, and we want to inform people as early as possible. I think that should be at the very top of the post. Let's make sure everyone knows as quickly as possible about what information they'll get from this post.

File: posts/2024-04-10-updates-to-rusts-wasi-targets.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+| 2024-09-05 | 1.81        | 1.82      | 1.83         | warn if `wasm32-wasi` is used on stable  |
+| 2024-10-17 | 1.82        | 1.83      | 1.84         | `wasm32-wasi` unavailable on nightly     |
+| 2024-11-28 | 1.83        | 1.84      | 1.85         | `wasm32-wasi` unavailable on beta        |
+| 2025-01-09 | 1.84        | 1.85      | 1.86         | `wasm32-wasi` unavailable on stable      |
+
+## Conclusion
+
+In this post we've discussed the upcoming updates to Rust's WASI targets. Come
+Rust 1.78 the `wasm32-wasip1` (tier 2) and `wasm32-wasip2` (tier 3) targets will
+be added. In Rust 1.81 we will begin warning if `wasm32-wasi` is being used. And in Rust 1.84, the existing `wasm32-wasi` target will be removed.
```
[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply posts/2024-04-10-updates-to-rusts-wasi-targets.md:97]: Good call!

File: posts/2024-04-10-updates-to-rusts-wasi-targets.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+knowing the target would likely undergo significant changes in the future. With
+the knowledge we have today though, we would not have chosen to introduce the
+"WASI, snapshot 1" target as `wasm32-wasi`. We should have instead chosen to add
+some suffix to the initial target triple so that the eventual stable WASI 1.0
+target can just be called `wasm32-wasi`.
+
+In anticipation of both an eventual WASI 1.0 target, and to preserve consistency
+between target names, we'll begin rolling out a name change to the existing WASI
+0.1 target. Starting in Rust 1.78 (May 2nd, 2024) a new `wasm32-wasip1` target
+will become available. Starting Rust 1.81 we will begin warning existing users
```
[the remark this answers is not in the document]
Comment (yoshuawuyts) [reply posts/2024-04-10-updates-to-rusts-wasi-targets.md:58]: Yes!

**Comment by yoshuawuyts** [comment]
@rylev Those are great points, thank you. Is this better?

**Review by alexcrichton — APPROVED** [review]

**Review by nikomatsakis — APPROVED** [review]
Exciting to see the progress!

**Comment by yoshuawuyts** [comment]
Reviews went through a lot quicker than expected - I think we're good to publish this today. Thanks everyone!

**Pull request merged**